Interface Material

lindsayad opened the pull request:
Create an `InterfaceMaterial` class capable of accessing face and neighboring face variable values and material properties

jiangwen84 committed:
make several changes from Material to MaterialBase, add tests
ref #12066

lindsayad committed:
Split out a MaterialBase class
ref #12066

Commit:
andrea material_base ref first commit #12066

Commit:
moving forward all test pass now

Commit:
almost there, still can't figure out why gentNeighborMaterialProperty doesn't work

Commit:
adding some override flags to comply with documentation errors

lindsayad committed:
Update to new MOOSE

lindsayad committed:
Make sure that neighbor subdomain id is set

lindsayad committed:
Dont treat InterfaceMaterials as boundary materials

lindsayad committed:
Separate warehouses for interface materials and non-interface materials

lindsayad committed:
Remove interface_material_data

lindsayad committed:
Remove reinit check

This check failed 10 tests. For one thing this check should have
no chance of working when looping over sides...after the first
side computation we would mark as not needing a reinit and then
the next side wouldn't have its material properties calculated

lindsayad committed:
Update tests to use MeshGenerators

lindsayad committed:
Interface value test

lindsayad committed:
Remove extraneous file

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/407261/) on ec4dbb6 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14124/site)

This comment will be updated on new commits.

lindsayad committed:
Don't reinit neighbor in ComputeElemAuxBcsThread

This would add a ghosting requirement for parallel computation

lindsayad committed:
Drop duplicated methods duplicated between base and derived

lindsayad committed:
Slacken neighbor check in ComputeMaterialsObjectThread::onInterface

This is actually necessary for any calculation with
neighbor stateful material properties. Interface objects
are identified with one unique side unlike InternalSide objects
which could be identified with either face. In the interface
case the elem_id < neighbor_id comparison is actually very bad.
If the element that has the face corresponding to the user specified
sideset has an ID greater than that of the neighboring element, then
the neighbor Material will never initialize its stateful properties

permcody commented:
Nice! - Oh goodie, looks like Rattlesnake gets broken again...

YaqiWang commented:
Happy to see that!

lindsayad commented:
Is anyone every gonna review this?

permcody commented:
I looked over this and didn't see any show stoppers. My comment should have been a little bit more specific. Are you going to make the patch for Rattlesnake or do you need help here? Perhaps @YaqiWang can help push this one over the finish line?

YaqiWang commented on test/src/materials/InterfaceValueMaterial.C:
```diff
@@ -0,0 +1,93 @@
+    _mp_master_name(getParam<std::string>("mat_prop_master")),
+    _mp_slave_name(getParam<std::string>("mat_prop_slave")),
+    _mp_master(getMaterialPropertyByName<Real>(_mp_master_name)),
+    _mp_slave(getNeighborMaterialProperty<Real>(_mp_slave_name)),
```
Why this one do not have `ByName` in it?

YaqiWang commented on test/src/materials/InterfaceValueMaterial.C:
```diff
@@ -0,0 +1,93 @@
+        getMaterialPropertyOld<Real>(_mp_out_base_name + "_" + std::string(_interface_value_type))),
+    _interface_value_2_old(getMaterialPropertyOld<Real>(_mp_var_out_base_name + "_" +
+                                                        std::string(_interface_value_type))),
+    _jump(declareProperty<Real>(static_cast<std::string>(_interface_value_type) + "_jump"))
```
We have to have static_cast here, or std::string itself is not enough?

lindsayad commented:
I will create a patch for Rattlesnake. Based on the error, it looks like it should be a very small fix. I'll make the fix now, and let you know when it's ready.

YaqiWang commented on test/src/materials/InterfaceValueMaterial.C:
```diff
@@ -0,0 +1,93 @@
+void
+InterfaceValueMaterial::computeQpProperties()
+{
+  mooseAssert(_mp_master[_qp] == _var_master[_qp],
```
Not obvious to me why we need these two assertions here.

lindsayad replied:
This was initially a `mooseWarning` created by @arovinelli; I changed it to a `mooseAssert`. He developed this test to make sure that he could feed in a non-zero neighbor var and a neighbor material property and make sure that they match (with something like `MaterialRealAux` involved somewhere)

arovinelli replied:
yep @lindsayad perfectly got the idea.
This was a check to make sure that using a material property or an aux variable would not affect the value of the jump. I agree with modifying it in an assert

YaqiWang commented on test/tests/interfacekernels/1d_interface/coupled_value_coupled_flux_with_jump_material.i:
```diff
@@ -0,0 +1,125 @@
+  xmax = 2
+[]
+
+[MeshModifiers]
```
I guess we are moving to the new mesh generation system.

YaqiWang commented on test/tests/interfacekernels/1d_interface/coupled_value_coupled_flux_with_jump_material.i:
```diff
@@ -0,0 +1,125 @@
+    variable = u
+    neighbor_var = v
+    boundary = master0_interface
+    penalty = 1e6
```
It makes more sense to me to change penalty to an interface material property. `jump_prop_name` makes this kernel too toylish.

lindsayad replied:
It is a toy though; it's in the test directory 😄

YaqiWang commented on test/tests/interfacekernels/1d_interface/tests:
```diff
@@ -18,6 +18,16 @@
+    design = 'InterfaceKernels/index.md'
+    issues = '#12066'
+    requirement = "The InterfaceKernel system shall use with interface material in 1D."
+                  "In this penalty implementation, jump is calculated by an interface material."
```
This line does not have to be in requirement.

YaqiWang commented on test/tests/interfacekernels/2d_interface/tests:
```diff
@@ -20,6 +20,17 @@
+    issues = "#12066"
+    design = "InterfaceKernels/index.md"
+    requirement = "The InterfaceKernel system shall use with interface material in 2D."
+                  "In this penalty implementation, jump is calculated by an interface material."
```
Same.

YaqiWang commented on test/tests/interfacekernels/3d_interface/coupled_value_coupled_flux_with_jump_material.i:
```diff
@@ -0,0 +1,154 @@
+[]
+
+[MeshModifiers]
+  [./subdomain1]
```
new mesh generation system.

YaqiWang commented on test/tests/interfacekernels/3d_interface/tests:
```diff
@@ -18,4 +18,14 @@
+    issues = "#12066"
+    design = "InterfaceKernels/index.md"
+    requirement = "The InterfaceKernel system shall use with interface material in 3D."
+                  "In this penalty implementation, jump is calculated by an interface material."
```
same

YaqiWang reviewed (commented):
Just browsed through this PR. Looks good to me. It is missing `InternalSideMaterial` but I guess we can merge this first and add that later. If we do so, I suggest that we create a new issue for adding `InternalSideMaterial`.

YaqiWang commented:
Thanks @lindsayad for making this happen. `InternalSideMaterial` is desired because it can be used to cleanup the penalty evaluations in Rattlesnake and remove several copies of DGKernels.

lindsayad commented:
> Thanks @lindsayad for making this happen. `InternalSideMaterial` is desired because it can be used to cleanup the penalty evaluations in Rattlesnake and remove several copies of DGKernels.

You're welcome. Cool, yea I will make sure to work on `InternalSideMaterial` next. After creating this, it should be pretty easy to follow this design pattern and create `InternalSideMaterial`.

lindsayad commented on framework/include/materials/InterfaceMaterial.h:
```diff
@@ -0,0 +1,239 @@
+
+  using MaterialBase::getZeroMaterialProperty;
+
+  virtual bool isBoundaryMaterial() const override { return _bnd; }
```
Indeed!

lindsayad committed:
Material cleanup

lindsayad committed:
Move mooseSetToZero to MathUtils

lindsayad committed:
Cleanup tests

lindsayad commented:
I have a fix ready for rattlesnake

lindsayad commented:
Thanks for your review @YaqiWang. I think I've addressed most things

arovinelli commented on framework/src/materials/InterfaceMaterial.C:
```diff
@@ -0,0 +1,52 @@
+    _current_elem(_assembly.elem()),
+    _neighbor_elem(_assembly.neighbor()),
+    _current_side(_assembly.side()),
+    _neighbor_side(_assembly.neighborSide())
```
you might want add the normal too,  but this is just a suggestion as I'don't know how much overhead there will be

arovinelli reviewed (commented):
@lindsayad @permcody @YaqiWang This is a pretty big change to MOOSE I'm wiring this as a comment but should really be a requirement. 
I might be wrong but I don't see any test using `getNeighborMatarialPropertyOld` (for the interface material) or the interface kernel using `getMatarialPropertyOld` (an old interface material property).
I might be overnparanoid but it might be something to add.
BTW @lindsayad thanks for taking the time to implement this, I really appreciate it!

lindsayad committed:
Add getNeighborMaterialPropertyByName

Commit:
added test for interface_material with broken mesh

Commit:
added test coupling old values

arovinelli commented:
@lindsayad I check out you branch and added a few tests making sure the new material can couple old variable values and material properties. You can find this changes in my [branch](https://github.com/arovinelli/moose/tree/material_base_alex) 
These changes should not break any other test
Thanks again for working on this

lindsayad commented:
Thanks @arovinelli I've incorporated those tests

arovinelli commented:
@lindsayad I forgot that when using `BreakMeshByBlockGenerator` coupled with `InterfaceKernel` we cannot recover. Also when using `BreakMeshByBlockGenerator` the mesh can only be replicated.
I update the tests in a new commit [here](https://github.com/arovinelli/moose/tree/material_base_alex)

arovinelli reviewed (approved):
@lindsayad @permcody @YaqiWang 
For me this is good to go

arovinelli commented:
@permcody @lindsayad @YaqiWang 
does this PR requires something else before being merged?

lindsayad commented:
I’m just waiting on review

> On Oct 22, 2019, at 9:24 AM, Andrea Rovinelli <notifications@github.com> wrote:
> 
> ﻿
> @permcody @lindsayad @YaqiWang
> does this PR requires something else before being merged?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

permcody commented:
I'll review it.

permcody commented:
@lindsayad thanks for pushing this over the finish line. I know it's been a long time coming and it was a lot of work. Excellent work. @arovinelli - thanks for the original contribution to get us down the down the path.

arovinelli commented:
@permcody and @alex,  thanks a lot for making this happen. It was a long way but the interface material is now a reality,
I'll finally be able to switch to current MOOSE  and add some CZM model to tensormechanics, such that everyone can use them.

Cheers and thanks again!

All the best!!

> @lindsayad thanks for pushing this over the finish line. I know it's been a long time coming and it was a lot of work. Excellent work. @arovinelli - thanks for the original contribution to get us down the down the path.
